### PR TITLE
feat: tlsping and tcpping using step-by-step

### PR DESCRIPTION
**Pull request by bassosimone**
## Checklist

- [x] I have read the [contribution guidelines](https://github.com/ooni/probe-cli/blob/master/CONTRIBUTING.md)
- [x] reference issue for this pull request: https://github.com/ooni/probe/issues/2158
- [x] if you changed anything related how experiments work and you need to reflect these changes in the ooni/spec repository, please link to the related ooni/spec pull request: https://github.com/ooni/spec/pull/250

## Description

This diff refactors the codebase to reimplement tlsping and tcpping
to use the step-by-step measurements style.

See docs/design/dd-003-step-by-step.md for more information on the
step-by-step measurement style.

### CODE REVIEW

File: internal/model/netx.go [hunk 1]
```diff
@@ -292,6 +292,68 @@ type TLSHandshaker interface {
 		net.Conn, tls.ConnectionState, error)
 }
 
+// Trace allows to collect measurement traces. A trace is injected into
+// netx operations using context.WithValue. Next code retrieves the trace
```
Comment (bassosimone) [inline internal/model/netx.go:296]:
```suggestion
// netx operations using context.WithValue. Netx code retrieves the trace
```

File: internal/netxlite/dialer.go [hunk 2]
```diff
@@ -171,15 +171,24 @@ func (d *DialerSystem) CloseIdleConnections() {
 	// nothing to do here
 }
 
-// dialerResolver combines dialing with domain name resolution.
-type dialerResolver struct {
+// dialerResolver combines dialing with domain name resolution and
```
Comment (bassosimone) [inline internal/netxlite/dialer.go:174]:
```suggestion
// dialerResolverWithTracing combines dialing with domain name resolution and
```

File: internal/netxlite/tls.go [hunk 3]
```diff
@@ -158,15 +158,15 @@ func NewTLSHandshakerStdlib(logger model.DebugLogger) model.TLSHandshaker {
...
-		DebugLogger: logger,
+		TLSHandshaker: th,
+		DebugLogger:   logger,
 	}
 }
 
 // tlsHandshakerConfigurable is a configurable TLS handshaker that
 // uses by default the standard library's TLS implementation.
+//
+// This type also implements events error wrapping and events tracing.
```
Comment (bassosimone) [inline internal/netxlite/tls.go:169]:
```suggestion
// This type also implements error wrapping and events tracing.
```

File: internal/netxlite/tls.go [hunk 4]
```diff
@@ -203,10 +214,19 @@ func (h *tlsHandshakerConfigurable) Handshake(
...
+	trace := ContextTraceOrDefault(ctx)
+	started := time.Now()
+	trace.OnTLSHandshakeStart(remoteAddr, config)
+	err = tlsconn.HandshakeContext(ctx)
+	err = maybeNewErrWrapper(classifyTLSHandshakeError, TLSHandshakeOperation, err)
+	finished := time.Now()
+	state := tlsMaybeConnectionState(tlsconn, err)
+	trace.OnTLSHandshakeDone(started, remoteAddr, config, state, err, finished)
+	if err != nil {
+		return nil, state, err
```
Comment (bassosimone) [inline internal/netxlite/tls.go:227]: We should test that the state we return here is always the zero state.

**Comment by DecFox** [comment]
Based on my conversation with @bassosimone, the main points from my initial review are:

1. The diff is easy to understand and complies with the _incremental changes_ approach we will be using to migrate to step-by-step style.
2. The diff essentially follows the proposed design (referenced [here](https://github.com/ooni/probe-cli/blob/master/docs/design/dd-003-step-by-step.md)).
3. We may document the use of buffered channels more clearly to make it easier to get acquainted with the new methodology.
4. Further review can be done once we have added sufficient tests.

File: internal/measurexlite/dialer_test.go [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+		}
+		events := trace.TCPConnects()
+		if len(events) != 0 {
+			t.Fatal("expected to see no TCPConnect events")
+		}
+	})
+}
+
+func TestArchivalSplitHostPort(t *testing.T) {
+	addr, port := archivalSplitHostPort("1.1.1.1")
```
Comment (bassosimone) [inline internal/measurexlite/dialer_test.go:185]:
```suggestion
	addr, port := archivalSplitHostPort("1.1.1.1") // missing port
```

File: internal/measurexlite/trace_test.go [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+			conn, err := dialer.DialContext(ctx, "tcp", "1.1.1.1:443")
+			if !errors.Is(err, mockedErr) {
+				t.Fatal("unexpected err", err)
+			}
+			if conn != nil {
+				t.Fatal("expected nil conn")
+			}
+		})
+
+		t.Run("with nonnil dependences and without hijacking", func(t *testing.T) {
```
Comment (bassosimone) [inline internal/measurexlite/trace_test.go:137]:
```suggestion
		t.Run("when nil", func(t *testing.T) {
```

File: internal/measurexlite/trace_test.go [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+				t.Fatal("unexpected err", err)
+			}
+			if conn != nil {
+				t.Fatal("expected nil conn")
+			}
+		})
+	})
+
+	t.Run("NewTLSHandshakerStdlibFn works as intended", func(t *testing.T) {
+		t.Run("with nonnil dependences and hijacking", func(t *testing.T) {
```
Comment (bassosimone) [inline internal/measurexlite/trace_test.go:155]:
```suggestion
		t.Run("when not nil", func(t *testing.T) {
```

File: internal/measurexlite/trace_test.go [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+			}
+			if !reflect.ValueOf(state).IsZero() {
+				t.Fatal("state is not a zero value")
+			}
+			if conn != nil {
+				t.Fatal("expected nil conn")
+			}
+		})
+
+		t.Run("with nonnil dependences and without hijacking", func(t *testing.T) {
```
Comment (bassosimone) [inline internal/measurexlite/trace_test.go:180]:
```suggestion
		t.Run("when nil", func(t *testing.T) {
```

**Review by bassosimone — COMMENTED** [review]
We still need to the following:

- [x] write tests ensuring that err wrapping is idempotent in terms of `errors.Is`: https://github.com/ooni/probe-cli/pull/815/commits/0d0e75a6e594c940723b81f9f59089191a07df49
- [x] write tests in `netxlite` to check whether we do tracing correctly for the `Dialer`: https://github.com/ooni/probe-cli/pull/815/commits/ea5e0b6d3d14dbb10523f96aeebb795e2727c13e
- [x] write tests in `netxlite` to check whether we do tracing correctly for the `TLSHandshake`: https://github.com/ooni/probe-cli/pull/815/commits/265fa607681a8c52c531e70bc886a8de085fad8c
- [x] write a test ensuring that `dialerResolverWithTracing.DialContext` wraps the returned `net.Conn` to implement error wrapping: https://github.com/ooni/probe-cli/pull/815/commits/ea5e0b6d3d14dbb10523f96aeebb795e2727c13e
- [x] write tests for `netxlite.MaybeNewErrWrapper`: https://github.com/ooni/probe-cli/pull/815/commits/0d0e75a6e594c940723b81f9f59089191a07df49
- [x] `tlsHandshakerConfigurable.Handshake` should have a test to ensure it returns the zero-value `tls.ConnectionState` on TLS handshake I/O error (or any other failure that could occur during the handshake): https://github.com/ooni/probe-cli/pull/815/commits/265fa607681a8c52c531e70bc886a8de085fad8c
- [x] write unit tests for `netxlite/trace.go`: https://github.com/ooni/probe-cli/pull/815/commits/f5dc906bd23759163f50a61c30f139917ee5f680
- [x] figure out why tests are not green: https://github.com/ooni/probe-cli/pull/815/commits/36db7fac3db02c790de010fcb33eccc7fb9ea059 (<- at least, this fixes the most frequent flakiness I saw)
- [x] write tests to ensure `netxlite` always gives wrapped errors to callbacks: we already have them
- [ ] create new issue for every new `TODO` introduced in this diff

**Comment by bassosimone** [comment]
Here's one of the test failures I have seen:

```
--- FAIL: TestTrace (0.00s)
    --- FAIL: TestTrace/NewTLSHandshakerStdlibFn_works_as_intended (0.00s)
        --- FAIL: TestTrace/NewTLSHandshakerStdlibFn_works_as_intended/when_nil (0.00s)
            trace_test.go:214: unexpected err unknown_failure: mocked
FAIL
coverage: 100.0% of statements
```

See https://github.com/ooni/probe-cli/runs/7020275622?check_suite_focus=true#step:4:165

File: internal/netxlite/tls.go (hunk 4 above)
Comment (bassosimone) [reply internal/netxlite/tls.go:227]: This is now tracked here: https://github.com/ooni/probe-cli/pull/815#pullrequestreview-1015249436

**Comment by DecFox** [comment]
As discussed with @bassosimone, the introduced tests cover all new functionalities in the step-by-step methodology. Tests are running successfully on linux and the flaky tests (while running inside CI) have been resolved. Also, the exposure of `NewErrWrapper` (`MaybeNewErrWrapper` now) in `netxlite` is a nice addition :)

**Pull request merged**